cocotb/decorators.py

```diff
@@ -336,20 +336,23 @@ def log(self):
     def __call__(self, *args, **kwargs):
 
         @coroutine
-        def execute_function(self, event):
+        def execute_function(self, event, back_event):
             coro = cocotb.coroutine(self._func)(*args, **kwargs)
             try:
                 _outcome = outcomes.Value((yield coro))
             except BaseException as e:
                 _outcome = outcomes.Error(e)
             event.outcome = _outcome
             event.set()
+            back_event.wait()
 
         event = threading.Event()
-        waiter = cocotb.scheduler.queue_function(execute_function(self, event))
+        back_event = threading.Event()
+        waiter = cocotb.scheduler.queue_function(execute_function(self, event, back_event))
         # This blocks the calling external thread until the coroutine finishes
         event.wait()
         waiter.thread_resume()
+        back_event.set()
         return event.outcome.get()
```

Comment:
A comment explaining the purpose of `back_event` would be good here. It seem the goal is to ensure that `thread_resume` is called before the `coroutine` completes. Can you explain _why_ that matters in the comment?